File: verifiers/python/src/deviceintelligence_verifier/keystream.py
"""v1 symmetric token crypto (Keystream.kt). Confidentiality in transit only."""
import hashlib

PHRASE = b"intel-verdict-token-key-v1"  # WIRE-CONSTANT (do NOT rebrand)


def decrypt_bytes(cipher: bytes) -> bytes:
    key = hashlib.sha256(PHRASE).digest()
    out = bytearray(len(cipher))
    off = block = 0
    while off < len(cipher):
        blk = block.to_bytes(4, "little")
        ks = hashlib.sha256(key + blk).digest()
        for i in range(min(32, len(cipher) - off)):
            out[off + i] = cipher[off + i] ^ ks[i]
        off += 32
        block += 1
    return bytes(out)


def decrypt_hex(token_hex: str) -> str:
    return decrypt_bytes(bytes.fromhex(token_hex.strip())).decode("utf-8")

File: verifiers/node/src/keystream.ts
// v1 symmetric token crypto (Keystream.kt port). Confidentiality in transit only.
import { createHash } from "node:crypto";

const PHRASE = "intel-verdict-token-key-v1"; // WIRE-CONSTANT (do NOT rebrand)

export function decryptBytes(cipher: Buffer): Buffer {
  const key = createHash("sha256").update(PHRASE, "utf8").digest();
  const out = Buffer.alloc(cipher.length);
  let off = 0, block = 0;
  while (off < cipher.length) {
    const blk = Buffer.alloc(4);
    blk.writeUInt32LE(block);
    const ks = createHash("sha256").update(key).update(blk).digest();
    for (let i = 0; i < Math.min(32, cipher.length - off); i++)
      out[off + i] = cipher[off + i] ^ ks[i];
    off += 32; block += 1;
  }
  return out;
}

export function decryptHex(tokenHex: string): string {
  return decryptBytes(Buffer.from(tokenHex.trim(), "hex")).toString("utf8");
}
